### ISSUE TITLE
section_path_distances gives an error for the hippocampus cell

### ISSUE THREAD

**Issue by chalimou** [BlueBrain/NeuroM#438]
The extraction of the feature section_path_distances for the hippocampus cell (included below) crashes with an error when computed with the following script: (this was not the case with ezy)

```
import neurom as nm

if __name__ == '__main__':
    ref_population = nm.load_neurons('./hippo')
    ref_data = nm.get('section_path_distances', ref_population, neurite_type=nm.AXON)
```
[hippo.zip](https://github.com/BlueBrain/NeuroM/files/363706/hippo.zip)

**Comment by juanchopanza** [comment]
I will look into it, but note that this neuron has two zero-length sections. This could be the root of the problem.

```
morph_check hippo/ca3b-N2.CNG.swc -d
```

Relevant part of output:

```
DEBUG: Nonzero section lengths: 2 failing ids detected: [1, 366]
ERROR:             Nonzero section lengths FAIL
```

Note I removed unnecessary stuff from your example.

**Comment by juanchopanza** [comment]
I think I found a potential issue: this file has two of its three soma points somewhere in the middle (SWC IDs 597 and 598). The SWC reader must be making some assumption that the somas are all at the beginning. I will create a small data sample to reproduce this and see about fixing it.

**Comment by juanchopanza** [comment]
OK, the morphology looks broken. Maybe @lida can comment on whether NeuroM should try to read a file with this structure or not. Here is the issue:

There is a starting soma point, that has two children: an apical dendrite branch, and another soma branch!

Here's the first soma point, with the first couple of basal dendrite points:
```
1 1 436.0 -1566.0 533.2 1.5 -1
2 3 435.1 -1583.0 500.0 1.5 1
3 3 438.2 -1586.0 500.0 1.5 2
```

And here is the second soma branch, which has two soma points and is followed by two apical dendrite points:
```
597 1 435.0 -1567.0 531.2 1.5 1
598 1 444.7 -1568.0 566.8 1.5 597
599 4 444.4 -1568.0 600.0 1.5 598
600 4 443.2 -1566.0 600.0 1.5 599
```

I don't think we can make sense out of this data. At the very least, it the soma points to not satisfy [any of out supported definitions](http://neurom.readthedocs.io/en/latest/definitions.html#soma), but even if we extend the number of definitions supported, I am not sure how we'd interpret these points.

**Comment by juanchopanza** [comment]
@chalimou Maybe you can consult with whoever created this file, but changing the type of points 597 and 598 to apical dendrite might be a solution.

**Comment by chalimou** [comment]
Thanks very much for the analysis Juan. I’ll check with Armando tomorrow and if we cannot find a solution, we’ll wait until Lida is back from vacation in three weeks time.
nancy

**Comment by juanchopanza** [comment]
@chalimou FYI, I took the liberty to reduce the example forward, so anyone reading it can see the most straight forward way to extract a feature.

**Comment by juanchopanza** [comment]
@chalimou Any news on this?

**Comment by chalimou** [comment]
Not yet, sorry, it will take another week I’m afraid since Armando is away next week. I sent him the link with the comments again yesterday but I do not think he had time to look at them.

thanks, Juan

**Comment by lidakanari** [comment]
@juanchopanza I just learned about this issue and what you describe is not considered an error in the morphology. According to the specs (http://research.mssm.edu/cnic/swc.html) there is no restriction about the ordering. Besides there was a long discussion a couple of months ago (about h5 format) and the outcome was that we should not assume any ordering.

Coming back to the example you describe, the soma points consist a single section. That means we cannot have a bifurcation within the soma. A tree starting from any point of the soma that is not the last one is not only allowed but is actually what we have in the majority of the swc data. 

So the following is correct:
```
1 1 x y z d -1
2 3 x y z d 1
3 3 x y z d 2
4 1 x y z d 1
```
and it is just a matter of ordering of the points.

But this is not correct:
```
1 1 x y z d -1
2 1 x y z d 1
3 3 x y z d 2
4 1 x y z d 1
```
because it assumes there is a bifurcation within the soma. 

From my understanding we have the first case, so the morphology is correct and just the ordering of the points is problematic. Please let me know if I misinterpreted something from the previous discussion.

**Comment by juanchopanza** [comment]
@lidakanari It isn't a matter of ordering, but one of topology + the special treatment of somas. This cell is a case where there are two branches coming out of the initial (soma) point. One is a soma, the other is a neurite. If that is a valid construction, then I may need to make changes to the code to accept it. Currently, I don't think it fits well in our definitions (the first point is a branching point. Is is a section in itself? Or is is the beginning of a section? We considered the first point to be  section in these cases to accommodate for single point somas, but this assumes no following points are soma points.)

BTW, note that the fix to issue #463 may make it easier to "fix" this if we decide we want to accept this kind of soma. But there could be other issues with this cell.

**Comment by lidakanari** [comment]
From what you report that is what I understand as the structure of the cells:

[test.pdf](https://github.com/BlueBrain/NeuroM/files/422469/test.pdf)

_If that is the case_:
1) We don't have 3 soma sections! It is one contour and trees attached to the right places
2) This is the actual and correct structure; once it is transformed into h5 it is actually not correct and we use it improperly in absence of a better representation. As a result we cannot use the approximation to this as our default since still many tools do not work with the one section soma and therefore other formats are necessary for the workflow!
3) This is a matter of ordering. If instead of figure 1 we had the following everything should be OK, otherwise the loader is absolutely wrong and it should be fixed in any case.

[test1.pdf](https://github.com/BlueBrain/NeuroM/files/422478/test1.pdf)

Are we on the same page in this? 

Now, if this is not the case, please forgive my ignorance and make a schematic of what the two section soma looks like. Then we can discuss if we should support or not this case.

**Comment by juanchopanza** [comment]
> From what you report that is what I understand as the structure of the cell:

No, as I mentioned, the first point has *two* branches, starting at points 2 and 597.

**Comment by lidakanari** [comment]
Draw a figure. I don't understand your interpretation, if it does not correspond to figure 1.

**Comment by juanchopanza** [comment]
You see how in your first figure point 1 only has one child, namely point 2? Well, in the actual file in question, point 1 has two children, points 2 and 597.

```
    1 --- 2 --- 3 --- 4 ....
    |
    | --- 597 --- 598 -- 599 ....
```

**Comment by lidakanari** [comment]
Sorry but this is not different from figure 1. Only point "1" is point "4".... OK, I will change the numbers for your comfort!

**Comment by juanchopanza** [comment]
Or maybe change the numbers so they actually relate to the data we're talking about.

**Comment by lidakanari** [comment]
Sorry, I didn't realize that I had to spend more time drawing the exact structure because of someone else's problem I did not cause. This is not a very professional approach but here you go. Is this ok now? 

[test2.pdf](https://github.com/BlueBrain/NeuroM/files/422511/test2.pdf)

If that is the case the file is correct, the loader is not. 99% of the swc files have this precise structure and as I said before the only point that we have to clarify is if we accept the arbitrary ordering. The trees are allowed to start from any soma point.

**Comment by juanchopanza** [comment]
#472 created for contour soma testing and #474 for somas of non-consecutive points.

**Comment by juanchopanza** [comment]
Here's a re-cap:

At the time this issue was reported, the kind of soma structure was not explicitly supported in NeuroM. The conclusion now is that it should be, given that is seems to be commonly used. Tests have been added to ensure that various previously unsupported soma configurations can be parsed by the SWC reader, and this seems to be the case. However, the error still occurs, which leads me to believe it is due to some other issue. I look at this in more detail when I come back from holiday.

**Comment by juanchopanza** [comment]
I think this is fixed by #478. The issue was with the soma being split into separate blocks in the file. The neuron builder was assuming all soma points are before neurite ones, and this lead to some neurites having soma sections as parents. And this changes the behaviour of the upstream iterator, which is used to calculate the path length.
